# 01 – Data Understanding: California Housing

Goal: understand the dataset before building any model.

## 1. Setup

### 1.1 Check the Python version
`assert` checks a rule. If the rule is false, the notebook stops with an error.
Here the rule is: Python must be 3.7 or newer.
This way an old Python fails now, at the start, instead of breaking code later.

In [1]:
import sys

assert sys.version_info >= (3, 7)

### 1.2 Check the scikit-learn version and import matplotlib
- `sklearn` = scikit-learn, the ML library we use to build models.
- Some functions we use only exist in version 1.0.1 or newer, so we check that here.
- `version.parse` turns a version text like "1.0.1" into something Python can compare correctly.
- `matplotlib.pyplot` (short name `plt`) is used to draw charts.

In [2]:
from packaging import version
import sklearn
import matplotlib.pyplot as plt

assert version.parse(sklearn.__version__) >= version.parse("1.0.1")

### 1.3 Import pandas and Path
- `pandas` (short name `pd`) works with tables of data, called DataFrames: rows and columns, like Excel.
- `Path` handles file and folder paths. It works the same on Windows and Mac/Linux,
  so the code also runs on GitHub's computer (CI). We use it later to save files.

In [3]:
import pandas as pd
from pathlib import Path

### 1.4 Import NumPy
- `numpy` (short name `np`) does fast maths on lists of numbers (arrays).
- pandas and scikit-learn are built on top of NumPy.
- We need it directly later, e.g. for gradient descent in notebook 02.

In [4]:
import numpy as np

## 2. Load the data

### 2.1 Get the California Housing dataset
- The data comes from scikit-learn. The first run downloads it once and saves it on the PC.
- `as_frame=True` gives it as a pandas table (DataFrame) instead of plain NumPy arrays.
- Each row = **one area** (neighbourhood) in California, not one house.
- Target `MedHouseVal` = the middle house price in that area, in **100,000 USD** (2.5 = $250,000).
- `head()` shows the first 5 rows.

In [5]:
from sklearn.datasets import fetch_california_housing

data = fetch_california_housing(as_frame=True)
df = data.frame
df.head()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


### 2.2 What the columns mean
| Column | Meaning |
|---|---|
| `MedInc` | middle income in the area (in 10,000 USD; 3.5 = $35,000) |
| `HouseAge` | middle age of the houses (years) |
| `AveRooms` | average rooms per home |
| `AveBedrms` | average bedrooms per home |
| `Population` | number of people living in the area |
| `AveOccup` | average people per home |
| `Latitude` | north–south position |
| `Longitude` | east–west position |
| `MedHouseVal` | **target**: middle house price (in 100,000 USD) |

The first 8 columns are the **features** (inputs). `MedHouseVal` is the **target** (what we predict).

### 2.3 Size, data types and missing values
`info()` shows:
- how many rows and columns there are
- the type of each column (`float64` = decimal number)
- the **Non-Null Count**: how many values are filled in. If it is lower than the number of rows, values are missing.

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   MedInc       20640 non-null  float64
 1   HouseAge     20640 non-null  float64
 2   AveRooms     20640 non-null  float64
 3   AveBedrms    20640 non-null  float64
 4   Population   20640 non-null  float64
 5   AveOccup     20640 non-null  float64
 6   Latitude     20640 non-null  float64
 7   Longitude    20640 non-null  float64
 8   MedHouseVal  20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB


**Findings:**
- Rows: Considering the above output there are 20640 entries or rows and   Columns: there are 9 Columns
- Missing values: There are no missing values
- Data types: and all data types are same float64